File: backend/src/loom/schemas/correlation.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class CorrelationCandidateMemberResponse(BaseModel):
    id: UUID
    asset_id: UUID
    original_filename: str | None = None
    capture_time: datetime | None = None

    model_config = {"from_attributes": True}


class CorrelationCandidateResponse(BaseModel):
    id: UUID
    case_id: UUID
    start_utc: datetime
    end_utc: datetime
    confidence: float = Field(ge=0.0, le=1.0)
    reasoning: dict[str, Any]
    status: Literal["pending", "accepted", "rejected"]
    decided_by: UUID | None = None
    decided_at: datetime | None = None
    members: list[CorrelationCandidateMemberResponse]
    created_at: datetime

    model_config = {"from_attributes": True}


class CorrelationCandidateListResponse(BaseModel):
    candidates: list[CorrelationCandidateResponse]
    total: int


class CorrelationCandidateDecisionRequest(BaseModel):
    """accept or reject a candidate. no merge happens here."""

    status: Literal["accepted", "rejected"]

File: frontend/src/types/correlation.ts
export type CorrelationStatus = 'pending' | 'accepted' | 'rejected';

export interface CorrelationCandidateMember {
  id: string;
  assetId: string;
  originalFilename: string | null;
  captureTime: string | null;
}

export interface CorrelationReasoning {
  [signal: string]:
    | {
        score: number | null;
        notes?: string | null;
      }
    | unknown;
}

export interface CorrelationCandidate {
  id: string;
  caseId: string;
  startUtc: string;
  endUtc: string;
  confidence: number;
  reasoning: CorrelationReasoning;
  status: CorrelationStatus;
  decidedBy: string | null;
  decidedAt: string | null;
  members: CorrelationCandidateMember[];
  createdAt: string;
}

export interface CorrelationCandidateListResponse {
  candidates: CorrelationCandidate[];
  total: number;
}

export interface CorrelationDecisionPayload {
  status: Exclude<CorrelationStatus, 'pending'>;
}
